### PR TITLE
Hot plug event handler modified to reduce cycles consumed

### PR DESCRIPTION
**Pull request by amuthuk1**
HotPlugEventHandler is in runtime path for all usecases and hence matters for performance.

Moved memset and fd query out of the loop
Reduced buffer size to 256
Added sleep to avoid busy looping

Test:Device boots to home screen
     Perf bench marks with AIA

Result: Improves performance.

Signed-off-by: Aravindan M <aravindan.muthukumar@intel.com>
Signed-off-by: Yogesh Marathe <yogesh.marathe@intel.com>

### CODE REVIEW

File: common/core/gpudevice.cpp [hunk 1]
```diff
@@ -201,6 +201,8 @@ void GpuDevice::DisplayManager::HotPlugEventHandler() {
           "UpdateDisplayState.");
       UpdateDisplayState();
     }
+    /* Adding sleep to avoid busy looping before a retry */
+    usleep(1000);
```
Comment (kalyankondapally) [inline common/core/gpudevice.cpp]: hmm we shouldnt busy loop here, we should return when we have nothing to read. Is sleep really needed here ?

Comment (yogesh-marathe) [inline common/core/gpudevice.cpp]: Kalyan, The loop is breaking / returning only when read fails. There could be a case of race / priority inversion (very corner case) where this thread will hog entire system, sleep ensures that wont happen. We should either remove the busy loop itself or have this sleep on the safer side. There is no perf improve due to this but this is more of precautionary measure. Busy loops are bad!

Comment (kalyankondapally) [inline common/core/gpudevice.cpp]: hmm we break when there is nothing more to read from fd. How can we ensure we are reading everything from fd without the loop ?

Comment (yogesh-marathe) [inline common/core/gpudevice.cpp]: Ok, in that case a loop is required so lets keep it with a sleep. We won't need sleep even if there was one blocking system call inside the loop but here that's not the case, that read() is non-blocking and will wait till it reads bytes or fails.

Comment (tpalli) [inline common/core/gpudevice.cpp]: I'm not familiar with this part of code but maybe check the size with lseek() and then read all that at once?

Comment (kalyankondapally) [inline common/core/gpudevice.cpp]: ya, probably we can split this sleep thing from this PR. In subsequent one, move to make use of lseek. I am afraid we will run into nasty race issues with sleep when u have "Real" hot plug events. Is that fine @yogesh-marathe

Comment (yogesh-marathe) [inline common/core/gpudevice.cpp]: I think lseek() will help when its really big buffer where you are reading in same buffer multiple times or in a file. Here I think its small number of bytes, which we read and parse until we find both strings (drm evnent and hot plug). If you add lseek here I think you are just adding one more system call.

Comment (kalyankondapally) [inline common/core/gpudevice.cpp]: Lets separate sleep to a separate PR, and we can check the perf impact, ensure it doesnt affect hotplug event in any way and make a call. Rest looks good to be merged.

Comment (yogesh-marathe) [inline common/core/gpudevice.cpp]: Yes we have checked hotplug functionality with this and that seems to be fine.

Comment (yogesh-marathe) [inline common/core/gpudevice.cpp]: Created PR without sleep.

Comment (tpalli) [inline common/core/gpudevice.cpp]: > If you add lseek here I think you are just adding one more system call.

IMO it would be more important to make the existing code more robust than worry about performance of a single system call. Adding sleeps to the code indicates that there is problem with the whole approach here.

Comment (yogesh-marathe) [inline common/core/gpudevice.cpp]: Tapani, Its not a single system call, HotPlugEventHandler gets called multiple times (even without actual hotplug, I didn't understand that behaviour!!), we have observed this when we put prints there. If lseek can get rid of loop itself we can get rid of that sleep at cost of 'one more sys call at runtime' (its not a single call).

**Comment by kalyankondapally** [comment]
merged
